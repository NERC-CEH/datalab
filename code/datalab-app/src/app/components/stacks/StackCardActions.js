import React from 'react';
import PropTypes from 'prop-types';
import CardActions from '@material-ui/core/CardActions';
import Button from '@material-ui/core/Button';
import PermissionWrapper from '../common/ComponentPermissionWrapper';
import { READY } from '../../../shared/statusTypes';

const StackCardActions = ({ stack, openStack, deleteStack, editStack, userPermissions, openPermission,
  deletePermission, editPermission }) => (
  <CardActions style={{ paddingLeft: 8, paddingRight: 8 }}>
    <PermissionWrapper userPermissions={userPermissions} permission={openPermission}>
      <Button style={{ marginRight: 4 }} color="primary" disabled={!openStack || !isReady(stack)} onClick={() => openStack(stack.id)}>
        Open
      </Button>
    </PermissionWrapper>
    <PermissionWrapper userPermissions={userPermissions} permission={deletePermission}>
      <Button style={{ marginLeft: 4, marginRight: 4 }} color="accent" disabled={!deleteStack || !isReady(stack)} onClick={() => deleteStack(stack)}>
        Delete
      </Button>
    </PermissionWrapper>
    <PermissionWrapper userPermissions={userPermissions} permission={editPermission}>
      <Button style={{ marginLeft: 4 }} color="accent" disabled={!editStack || !isReady(stack)} onClick={() => editStack(stack)}>
        Edit
      </Button>
    </PermissionWrapper>
  </CardActions>
);

StackCardActions.propTypes = {
  stack: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
    type: PropTypes.string,
    status: PropTypes.string,
  }).isRequired,
  openStack: PropTypes.func,
  deleteStack: PropTypes.func,
  editStack: PropTypes.func,
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
  openPermission: PropTypes.string.isRequired,
  deletePermission: PropTypes.string.isRequired,
  editPermission: PropTypes.string.isRequired,
};

const isReady = ({ status }) => status === READY;

export default StackCardActions;
