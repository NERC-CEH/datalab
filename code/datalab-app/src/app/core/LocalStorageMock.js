class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key];
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export default LocalStorageMock;
