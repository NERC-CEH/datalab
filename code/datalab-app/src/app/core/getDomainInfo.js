function extendSubdomain(extension, localhostPort, location) {
  const domainInfo = getDomainInfo(location);

  if (domainInfo === 'localhost') {
    return `http://localhost:${localhostPort}`;
  }

  return `${domainInfo.protocol}//${domainInfo.subdomain}-${extension}.${domainInfo.domain}`;
}

function replaceSubdomain(subdomain, altHref, location) {
  const domainInfo = getDomainInfo(location);

  if (domainInfo === 'localhost') {
    return altHref || 'localhost';
  }

  return `${domainInfo.protocol}//${subdomain}.${domainInfo.domain}`;
}

function getDomainInfo(location) {
  const currentLocation = location || window.location;

  if (currentLocation.hostname === 'localhost') {
    return 'localhost';
  }

  const { protocol } = currentLocation;
  const { hostname } = currentLocation;
  const subdomain = hostname.split('.')[0];
  const domain = hostname.substring(subdomain.length + 1, hostname.length);

  return { protocol, subdomain, domain };
}

export { getDomainInfo, replaceSubdomain, extendSubdomain };
