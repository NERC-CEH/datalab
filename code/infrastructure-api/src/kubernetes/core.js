import logger from 'winston';
import { has, get } from 'lodash';

export const handleCreateError = (type, name) => (error) => {
  if (has(error, 'response.data.message')) {
    throw createError(`Unable to create kubernetes ${type} '${name}' - ${error.response.data.message}`);
  }
  throw createError(`unable to create kubernetes ${type} '${name}' - ${error}`);
};

export const handleDeleteError = (type, name) => (error) => {
  if (has(error, 'response.status') && get(error, 'response.status') === 404) {
    logger.warn(`Kubernetes API: Could not find ${type}: ${name} to delete it`);
    return Promise.resolve();
  }

  logger.error(`Error deleting ${type}: ${name}`);
  throw createError(error.message);
};

function createError(message) {
  return new Error(`Kubernetes API: ${message}`);
}
