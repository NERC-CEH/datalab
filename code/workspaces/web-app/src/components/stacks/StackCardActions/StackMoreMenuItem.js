import React from 'react';
import PropTypes from 'prop-types';
import MenuItem from '@material-ui/core/MenuItem';
import Tooltip from '@material-ui/core/Tooltip';
import PermissionWrapper from '../../common/ComponentPermissionWrapper';

// Uses forwardRef to forward ref from menu onto MenuItem as functional components cannot be given refs.
// This stops React logging an error and seems like the right place to forward the ref too as
// this component is a wrapper around the MenuItem.
const StackMoreMenuItem = React.forwardRef((props, ref) => {
  const {
    shouldRender = false, onClick, disabled, userPermissions, requiredPermission,
    tooltipText = '', disableTooltip, children,
  } = props;

  if (!shouldRender) return null;

  return (
    <PermissionWrapper userPermissions={userPermissions} permission={requiredPermission}>
      <Tooltip
        title={tooltipText}
        disableHoverListener={disableTooltip}>
        <div> {/* provides element for tooltip to listen to when menu item is disabled */}
          <MenuItem disabled={disabled} onClick={onClick} ref={ref}>
            {children}
          </MenuItem>
        </div>
      </Tooltip>
    </PermissionWrapper>
  );
});

StackMoreMenuItem.propTypes = {
  shouldRender: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
  requiredPermission: PropTypes.string.isRequired,
  tooltipText: PropTypes.string,
  disableTooltip: PropTypes.bool,
  children: PropTypes.node,
};

export default StackMoreMenuItem;
