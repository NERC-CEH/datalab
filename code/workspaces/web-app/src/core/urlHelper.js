export default function getUrlNameStartEndText(projectKey, windowLocation) {
  const separator = '.';
  const restHostname = windowLocation.hostname.split(separator).slice(1);
  const startText = `${windowLocation.protocol}//${projectKey}-`;

  let endText = `${separator}${restHostname.join(separator)}`;

  if (windowLocation.hostname === 'localhost') {
    endText = '.datalabs.localhost';
  }

  return { startText, endText };
}
