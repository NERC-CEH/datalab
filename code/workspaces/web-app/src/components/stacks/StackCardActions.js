import { withStyles } from '@material-ui/core/styles';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Icon from '@material-ui/core/Icon';
import PropTypes from 'prop-types';
import React from 'react';
import { statusTypes } from 'common';
import { useCurrentUserId } from '../../hooks/authHooks';
import PermissionWrapper from '../common/ComponentPermissionWrapper';
import PrimaryActionButton from '../common/buttons/PrimaryActionButton';
import SecondaryActionButton from '../common/buttons/SecondaryActionButton';

const { READY } = statusTypes;
const MORE_ICON = 'more_vert';

const styles = theme => ({
  cardActions: {
    display: 'flex',
    width: '100%',
  },
  buttonWrapper: {
    flexGrow: '1',
    '& + &': {
      marginLeft: theme.spacing(1),
    },
  },
});

const StackCardActions = (props) => {
  const currentUserId = useCurrentUserId();
  return <PureStackCardActions currentUserId={currentUserId} {...props} />;
};

export const PureStackCardActions = ({ stack, openStack, deleteStack, editStack, userPermissions, openPermission,
  deletePermission, editPermission, currentUserId, classes }) => {
  // Treat user as owner if 'users' not a defined field on stack.
  // This is the case for projects which also use this component. This will mean that
  // projects rely solely on the permissions passed to determine correct rendering.
  const ownsStack = !stack.users || stack.users.includes(currentUserId);
  const [anchorEl, setAnchorEl] = React.useState(null);

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div className={classes.cardActions}>
      {openStack && <PermissionWrapper className={classes.buttonWrapper} userPermissions={userPermissions} permission={openPermission}>
        <PrimaryActionButton
          disabled={!isReady(stack)}
          onClick={() => openStack(stack)}
          fullWidth
        >
          Open
        </PrimaryActionButton>
      </PermissionWrapper>}
      {ownsStack && <PermissionWrapper className={classes.buttonWrapper} userPermissions={userPermissions} permission={deletePermission}>
        <SecondaryActionButton
          disabled={!isReady(stack)}
          aria-controls="more-menu"
          aria-haspopup="true"
          onClick={handleClick}
        >
          <Icon style={{ color: 'inherit' }}>{MORE_ICON}</Icon>
        </SecondaryActionButton>
      </PermissionWrapper>}
      <Menu
        id="more-menu"
        anchorEl={anchorEl}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        {editStack && ownsStack && <PermissionWrapper userPermissions={userPermissions} permission={editPermission}>
          <MenuItem
            disabled={!isReady(stack)}
            onClick={() => editStack(stack)}
          >
            Edit
          </MenuItem>
        </PermissionWrapper>}
        {deleteStack && ownsStack && <PermissionWrapper userPermissions={userPermissions} permission={deletePermission}>
          <MenuItem
            disabled={!isReady(stack)}
            onClick={() => deleteStack(stack)}
          >
            Delete
          </MenuItem>
        </PermissionWrapper>}
      </Menu>
    </div>
  );
};

StackCardActions.propTypes = {
  stack: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
    type: PropTypes.string,
    status: PropTypes.string,
  }).isRequired,
  openStack: PropTypes.func,
  deleteStack: PropTypes.func,
  editStack: PropTypes.func,
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
  openPermission: PropTypes.string.isRequired,
  deletePermission: PropTypes.string.isRequired,
  editPermission: PropTypes.string.isRequired,
};

const isReady = ({ status }) => status === READY;

export default withStyles(styles)(StackCardActions);
