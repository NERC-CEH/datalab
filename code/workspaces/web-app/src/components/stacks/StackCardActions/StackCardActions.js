// @ts-nocheck
import withStyles from '@mui/styles/withStyles';
import Menu from '@mui/material/Menu';
import Icon from '@mui/material/Icon';
import Tooltip from '@mui/material/Tooltip';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import { statusTypes, visibilityTypes } from 'common';
import { SYSTEM_INSTANCE_ADMIN } from 'common/src/permissionTypes';
import { SITE_CATEGORY } from 'common/src/config/images';
import { useCurrentUserId } from '../../../hooks/authHooks';
import PermissionWrapper from '../../common/ComponentPermissionWrapper';
import PrimaryActionButton from '../../common/buttons/PrimaryActionButton';
import SecondaryActionButton from '../../common/buttons/SecondaryActionButton';
import StackMoreMenuItem from './StackMoreMenuItem';
import { getUserActionsForType } from '../../../config/images';

const { READY, SUSPENDED } = statusTypes;
const { PRIVATE, PROJECT, PUBLIC } = visibilityTypes;
const MORE_ICON = 'more_vert';

const styles = theme => ({
  cardActions: {
    display: 'flex',
    width: '100%',
  },
  buttonWrapper: {
    flexGrow: '1',
    '& + &': {
      marginLeft: theme.spacing(1),
    },
  },
});

const StackCardActions = (props) => {
  const { stack } = props;
  const currentUserId = useCurrentUserId();

  return <PureStackCardActions
    currentUserId={currentUserId}
    userActions={getUserActionsForType(stack.type)}
    {...props}
  />;
};

const getShareButton = (stack, shareStack, permission) => (newStatus, label, disabled) => ({
  onClick: () => shareStack(stack, newStatus),
  requiredPermission: permission,
  name: label,
  disabled,
  tooltipText: `Access already set to ${newStatus}`,
  disableTooltip: !disabled,
});

export const getSharedButtons = (stack, shareStack, permission) => {
  // Function to get the different "Share" buttons available to the stack
  const isSite = stack.category === SITE_CATEGORY;
  const shareButton = getShareButton(stack, shareStack, permission);

  const shared = stack.shared || stack.visible;

  return [
    isSite ? shareButton(PRIVATE, 'Set Access: Private', shared === PRIVATE) : undefined,
    shareButton(PROJECT, 'Set Access: Project', shared === PROJECT),
    isSite ? shareButton(PUBLIC, 'Set Access: Public', shared === PUBLIC) : undefined,
  ];
};

export const PureStackCardActions = ({ stack, openStack, deleteStack, editStack, restartStack, scaleStack, copySnippets, userActions,
  userPermissions, openPermission, deletePermission, editPermission, scalePermission, currentUserId, classes, getLogs, shareStack }) => {
  // Treat user as owner if 'users' not a defined field on stack.
  // This is the case for projects which also use this component. This will mean that
  // projects rely solely on the permissions passed to determine correct rendering.
  const ownsStack = !stack.users || stack.users.includes(currentUserId);
  const isInstanceAdmin = userPermissions.includes(SYSTEM_INSTANCE_ADMIN);
  const [anchorEl, setAnchorEl] = useState(null);

  const handleMoreButtonClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const handleMoreMenuClose = () => {
    setAnchorEl(null);
  };

  const OpenButton = React.forwardRef((props, ref) => <PrimaryActionButton ref={ref} {...props} />);

  const shouldRenderLogs = userActions.logs && getLogs !== undefined;
  const shouldRenderEdit = userActions.edit && editStack !== undefined;
  const shouldRenderShare = userActions.share && shareStack !== undefined;
  const shouldRenderRestart = userActions.restart && restartStack !== undefined;
  const shouldRenderScale = userActions.scale && scaleStack !== undefined;
  const shouldRenderDelete = userActions.delete && deleteStack !== undefined;
  const shouldRenderCopySnippet = userActions.copySnippets && copySnippets !== undefined;

  const copySnippetMenuItems = shouldRenderCopySnippet ? Object.keys(copySnippets).map(k => ({
    onClick: () => { handleMoreMenuClose(); copySnippets[k](stack); },
    requiredPermission: openPermission,
    name: `Copy ${k} snippet`,
  })) : [];

  const shareMenuItems = shouldRenderShare ? getSharedButtons(stack, shareStack, deletePermission) : [];
  const scaleOption = stack.status === SUSPENDED ? 1 : 0;

  const stackMenuItems = [
    shouldRenderLogs && { onClick: () => getLogs(stack), requiredPermission: deletePermission, name: 'Logs' },
    shouldRenderEdit && { onClick: () => editStack(stack), requiredPermission: editPermission, name: 'Edit' },
    ...shareMenuItems,
    shouldRenderRestart && {
      onClick: () => restartStack(stack),
      requiredPermission: editPermission,
      name: 'Restart',
      disabled: stack.status === SUSPENDED,
    },
    shouldRenderScale && {
      onClick: () => scaleStack(stack, scaleOption),
      requiredPermission: scalePermission,
      name: stack.status === SUSPENDED ? 'Turn On' : 'Suspend',
    },
    ...copySnippetMenuItems,
    shouldRenderDelete && {
      onClick: () => deleteStack(stack),
      requiredPermission: deletePermission,
      name: 'Delete',
    },
  ].filter(m => !!m);

  const shouldRenderMenuItems = (ownsStack || isInstanceAdmin) && stackMenuItems.length > 0;

  return (
    <div className={classes.cardActions}>
      {openStack && <PermissionWrapper className={classes.buttonWrapper} userPermissions={userPermissions} permission={openPermission}>
        <Tooltip
          title='Cannot be opened until resource is ready' placement='bottom-start'
          disableHoverListener={isReady(stack)}
        >
          <div>
            {stack.status !== SUSPENDED
            && <OpenButton
              disabled={!isReady(stack)}
              onClick={() => openStack(stack)}
              fullWidth
            >
              Open
            </OpenButton>
            }
            {stack.status === SUSPENDED
            && <SecondaryActionButton
              onClick={() => scaleStack(stack, scaleOption)}
              fullWidth
            >
              Turn On
            </SecondaryActionButton>
            }
          </div>
        </Tooltip>
      </PermissionWrapper>}
      {shouldRenderMenuItems && <PermissionWrapper className={classes.buttonWrapper} userPermissions={userPermissions} permission={deletePermission}>
        <SecondaryActionButton
          aria-controls="more-menu"
          aria-haspopup="true"
          onClick={handleMoreButtonClick}
          fullWidth
        >
          <Icon style={{ color: 'inherit' }}>{MORE_ICON}</Icon>
        </SecondaryActionButton>
      <Menu
        id="more-menu"
        anchorEl={anchorEl}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleMoreMenuClose}
      >
        {stackMenuItems.map(m => (<StackMoreMenuItem
            shouldRender={true}
            userPermissions={userPermissions}
            {...m}
            key={m.name}
          >
            {m.name}
          </StackMoreMenuItem>))}
      </Menu>
      </PermissionWrapper>}
    </div>
  );
};

const sharedPropTypes = {
  stack: PropTypes.shape({
    id: PropTypes.string,
    displayName: PropTypes.string,
    type: PropTypes.string,
    category: PropTypes.string,
    status: PropTypes.string,
  }).isRequired,
  openStack: PropTypes.func,
  deleteStack: PropTypes.func,
  editStack: PropTypes.func,
  restartStack: PropTypes.func,
  scaleStack: PropTypes.func,
  getLogs: PropTypes.func,
  shareStack: PropTypes.func,
  copySnippets: PropTypes.objectOf(PropTypes.func),
  userPermissions: PropTypes.arrayOf(PropTypes.string).isRequired,
  openPermission: PropTypes.string,
  deletePermission: PropTypes.string,
  editPermission: PropTypes.string,
  scalePermission: PropTypes.string,
};

StackCardActions.propTypes = sharedPropTypes;
PureStackCardActions.propTypes = {
  ...sharedPropTypes,
  currentUserId: PropTypes.string.isRequired,
  userActions: PropTypes.shape({
    share: PropTypes.bool,
    edit: PropTypes.bool,
    restart: PropTypes.bool,
    delete: PropTypes.bool,
    logs: PropTypes.bool,
    copySnippets: PropTypes.bool,
  }).isRequired,
};

const isReady = ({ status }) => status === READY;

export default withStyles(styles)(StackCardActions);
