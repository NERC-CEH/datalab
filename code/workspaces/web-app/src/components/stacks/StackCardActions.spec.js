import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import StackCardActions from './StackCardActions';

describe('StackCardActions', () => {
  let shallow;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
  });

  function shallowRender(props) {
    return shallow(<StackCardActions {...props} />);
  }

  const openStackMock = jest.fn();
  const deleteStackMock = jest.fn();
  const editStackMock = jest.fn();

  const generateProps = () => ({
    stack: {
      id: 'abc1234',
      displayName: 'expectedDisplayName',
      type: 'expectedType',
      status: 'ready',
    },
    openStack: openStackMock,
    deleteStack: deleteStackMock,
    editStack: editStackMock,
    userPermissions: ['open', 'delete', 'edit'],
    openPermission: 'open',
    deletePermission: 'delete',
    editPermission: 'edit',
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('creates correct snapshot', () => {
    // Arrange
    const props = generateProps();

    // Act
    const output = shallowRender(props);

    // Assert
    expect(output).toMatchSnapshot();
  });

  it('Open button onClick function calls openStack with correct props', () => {
    // Arrange
    const props = generateProps();

    // Act
    const output = shallowRender(props);
    const onClick = output.find({ children: 'Open' }).prop('onClick');

    // Assert
    expect(openStackMock).not.toHaveBeenCalled();
    onClick();
    expect(openStackMock).toHaveBeenCalledTimes(1);
    expect(openStackMock).toHaveBeenCalledWith({
      displayName: 'expectedDisplayName',
      id: 'abc1234',
      type: 'expectedType',
      status: 'ready',
    });
  });

  it('Delete button onClick function calls openStack with correct props', () => {
    // Arrange
    const props = generateProps();

    // Act
    const output = shallowRender(props);
    const onClick = output.find({ children: 'Delete' }).prop('onClick');

    // Assert
    expect(deleteStackMock).not.toHaveBeenCalled();
    onClick();
    expect(deleteStackMock).toHaveBeenCalledTimes(1);
    expect(deleteStackMock).toHaveBeenCalledWith({
      displayName: 'expectedDisplayName',
      id: 'abc1234',
      type: 'expectedType',
      status: 'ready',
    });
  });

  it('Edit button onClick function calls openStack with correct props', () => {
    // Arrange
    const props = generateProps();

    // Act
    const output = shallowRender(props);
    const onClick = output.find({ children: 'Edit' }).prop('onClick');

    // Assert
    expect(editStackMock).not.toHaveBeenCalled();
    onClick();
    expect(editStackMock).toHaveBeenCalledTimes(1);
    expect(editStackMock).toHaveBeenCalledWith({
      displayName: 'expectedDisplayName',
      id: 'abc1234',
      type: 'expectedType',
      status: 'ready',
    });
  });
});
