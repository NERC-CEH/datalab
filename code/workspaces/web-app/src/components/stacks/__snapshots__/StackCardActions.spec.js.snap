// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackCardActions creates correct snapshot 1`] = `
<WithStyles(ForwardRef(CardActions))
  style={
    Object {
      "justifyContent": "space-between",
      "marginTop": 4,
      "padding": 0,
    }
  }
>
  <ComponentWrapper
    permission="open"
    userPermissions={
      Array [
        "open",
        "delete",
        "edit",
      ]
    }
  >
    <WithStyles(ForwardRef(Button))
      color="primary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": 4,
          "marginRight": 4,
        }
      }
      variant="contained"
    >
      Open
    </WithStyles(ForwardRef(Button))>
  </ComponentWrapper>
  <ComponentWrapper
    permission="delete"
    userPermissions={
      Array [
        "open",
        "delete",
        "edit",
      ]
    }
  >
    <WithStyles(ForwardRef(Button))
      color="secondary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": 4,
          "marginRight": 4,
        }
      }
      variant="contained"
    >
      Delete
    </WithStyles(ForwardRef(Button))>
  </ComponentWrapper>
  <ComponentWrapper
    permission="edit"
    userPermissions={
      Array [
        "open",
        "delete",
        "edit",
      ]
    }
  >
    <WithStyles(ForwardRef(Button))
      color="secondary"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": 4,
          "marginRight": 4,
        }
      }
      variant="contained"
    >
      Edit
    </WithStyles(ForwardRef(Button))>
  </ComponentWrapper>
</WithStyles(ForwardRef(CardActions))>
`;
